File: libgomp/testsuite/libgomp.fortran/declare-mapper-9.f90
! { dg-do run }

type t
  integer, dimension (8) :: arr1
end type t
type u
  type(t), dimension (:), pointer :: tarr
end type u

type(u) :: myu
type(t), dimension (1), target :: myarray

!$omp declare mapper (named: t :: x) map(x%arr1(1:4))
!$omp declare mapper (u :: x) map(to: x%tarr) map(mapper(named), tofrom: x%tarr(1))

myu%tarr => myarray
myu%tarr(1)%arr1 = 0

! Unnamed mapper invoking named mapper

!$omp target
myu%tarr(1)%arr1(1) = myu%tarr(1)%arr1(1) + 1
!$omp end target

if (myu%tarr(1)%arr1(1).ne.1) stop 1

end

File: libgomp/testsuite/libgomp.c-c++-common/declare-mapper-9.c
/* { dg-do run { target c++ } } */

#include <string.h>
#include <stdlib.h>
#include <assert.h>

#define N 64

struct A {
  int *arr1;
  float *arr2;
  int arr3[N];
};

int
main (int argc, char *argv[])
{
  struct A var;

  memset (&var, 0, sizeof var);
  var.arr1 = (int *) calloc (N, sizeof (int));
  var.arr2 = (float *) calloc (N, sizeof (float));

  {
    #pragma omp declare mapper (struct A x) map(to: x.arr1) \
					    map(tofrom: x.arr1[0:N])
    #pragma omp target
    {
      for (int i = 0; i < N; i++)
	var.arr1[i]++;
    }
  }

  {
    #pragma omp declare mapper (struct A x) map(to: x.arr2) \
					    map(tofrom: x.arr2[0:N])
    #pragma omp target
    {
      for (int i = 0; i < N; i++)
	var.arr2[i]++;
    }
  }

  {
    #pragma omp declare mapper (struct A x) map(tofrom: x.arr3[0:N])
    #pragma omp target
    {
      for (int i = 0; i < N; i++)
	var.arr3[i]++;
    }
  }

  for (int i = 0; i < N; i++)
    {
      assert (var.arr1[i] == 1);
      assert (var.arr2[i] == 1);
      assert (var.arr3[i] == 1);
    }

  free (var.arr1);
  free (var.arr2);
}
